blockchain/Validation.go

```diff
@@ -2805,7 +2827,50 @@ func IsCurrentForFeeEstimation() bool {
 func CheckInputsFromMempoolAndCache(tx *model.Tx, state *model.ValidationState, cache *utxo.CoinsViewCache,
 	mpool *mempool.Mempool, flags uint32, cacheSigStore bool, txData *model.PrecomputedTransactionData) bool {
 
-	return true
+	// todo AssertLockHeld(cs_main)
+	// pool.cs should be locked already, but go ahead and re-take the lock here
+	// to enforce that mempool doesn't change between when we check the view and
+	// when we actually call through to CheckInputs
+	mpool.Mtx.Lock()
+	defer mpool.Mtx.Unlock()
+
+	if tx.IsCoinBase() {
+		panic("critical error")
+	}
+	for _, txin := range tx.Ins {
+		coin := cache.AccessCoin(txin.PreviousOutPoint)
+
+		// At this point we haven't actually checked if the coins are all
+		// available (or shouldn't assume we have, since CheckInputs does). So
+		// we just return failure if the inputs are not available here, and then
+		// only have to check equivalence for available inputs.
+		if coin.IsSpent() {
+			return false
+		}
+
+		txFrom := mpool.Get(&txin.PreviousOutPoint.Hash)
+		if txFrom != nil {
+			if txFrom.TxHash() != txin.PreviousOutPoint.Hash {
+				panic("critical error")
+			}
+			if len(txFrom.Outs) <= int(txin.PreviousOutPoint.Index) {
+				panic("critical error")
+			}
+			if txFrom.Outs[txin.PreviousOutPoint.Index] != coin.TxOut {
```

Comment:
I think you can add IsEuqal function to compare the two TxOut .

Reply:
yes